src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -1838,10 +1838,23 @@ class RGWDataFullSyncShardCR : public RGWDataBaseSyncShardCR {
 				 error_repo, entry_timestamp, lease_cr,
 				 bucket_shard_cache, &*marker_tracker, tn),
 			       sc->lcc.adj_concurrency(cct->_conf->rgw_data_sync_spawn_window),
-			       std::nullopt);
+			       [&](uint64_t stack_id, int ret) {
+                                if (ret < 0) {
+                                  retcode = ret;
+                                }
+                                return retcode;
+                                });
           }
-	  sync_marker.marker = iter->first;
+          sync_marker.marker = iter->first;
         }
+
+        drain_all_cb([&](uint64_t stack_id, int ret) {
+                if (ret < 0) {
+                  retcode = ret;
+                }
+                return retcode;
+              });
```

Comment:
do we really need to drain inside this do-while loop? as long as there are more entries to process, we should try to keep the spawn window full. we do need a drain before calling `marker_tracker->flush()`, but that probably belongs just below this do-while loop

Reply:
yeah there was a drain call here after the do-while loop that I had removed as part of the first commit.  I guess just adding it back should be enough.

Reply:
thinking more about it, I remember we had to add similar error handling in incremental sync because we were failing to propagate the error from marker_tracker update failures from data_sync_single_entry(). I think we need this here as well in full sync.

Reply:
@cbodley do you still have an objection to this? we added https://github.com/ceph/ceph/commit/fb92311b3e17101c2eee4bc0e4ef7cf3b744bc7a to return error from RGWLastCallerWinsCR() if it should fail.

Reply:
wouldn't the `yield_spawn_window()` callback get that error without us needing an extra drain here?

Reply:
sorry I was only focusing on the error handling. We don't need to drain inside the loop. Removed.